File: 이것이 취업을 위한 코딩 테스트다/정렬/국영수/국영수.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

struct Student {
  string name;
  int kor;
  int eng;
  int m;

  Student(string name, int kor, int eng, int m) {
    this->name = name;
    this->kor = kor;
    this->eng = eng;
    this->m = m;
  }

  bool operator < (Student &other) {
    if (this->kor == other.kor && this->eng == other.eng && this->m == other.m) {
      return this->name < other.name;
    }
    if (this->kor == other.kor && this->eng == other.eng) {
      return this->m > other.m;
    }
    if (this->kor == other.kor) {
      return this->eng < other.eng;
    }
    return this->kor > other.kor;
  }
};

int n;
vector<Student> v;

int main(void) {
  cin >> n;
  for (int i = 0; i < n; i += 1) {
    string name;
    int kor;
    int eng;
    int m;
    cin >> name >> kor >> eng >> m;
    v.push_back(Student(name, kor, eng, m));
  }

  sort(v.begin(), v.end());

  for (int i = 0; i < n; i += 1) {
    cout << v[i].name << endl;
  }
}

File: 이것이 취업을 위한 코딩 테스트다/정렬/국영수/국영수.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Scanner;

public class 국영수 {
  static class Student implements Comparable<Student> {
    public String name;
    public int kor;
    public int eng;
    public int m;

    Student(String name, int kor, int eng, int m) {
      this.name = name;
      this.kor = kor;
      this.eng = eng;
      this.m = m;
    }

    @Override
    public int compareTo(국영수.Student other) {
      if (this.kor == other.kor && this.eng == other.eng && this.m == other.m) {
        return this.name.compareTo(other.name);
      }
      if (this.kor == other.kor && this.eng == other.eng) {
        return Integer.compare(other.m, this.m);
      }
      if (this.kor == other.kor) {
        return Integer.compare(this.eng, other.eng);
      }
      return Integer.compare(other.kor, this.kor);
    }
  }

  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    int n = sc.nextInt();

    ArrayList<Student> students = new ArrayList<>();
    for (int i = 0; i < n; i += 1) {
      String name = sc.next();
      int kor = sc.nextInt();
      int eng = sc.nextInt();
      int m = sc.nextInt();
      students.add(new Student(name, kor, eng, m));
    }

    Collections.sort(students);

    for (int i = 0; i < n; i += 1) {
      System.out.println(students.get(i).name);
    }
  }
}
